raft/quorum/quorum.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2019 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package quorum
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+)
+
+// IndexLookuper allows looking up a commit index for a given ID of a voter
+// from a corresponding MajorityConfig.
+type IndexLookuper interface {
+	Index(voterID uint64) (idx uint64, found bool)
+}
+
+type mapLookuper map[uint64]uint64
+
+func (m mapLookuper) Index(id uint64) (uint64, bool) {
+	idx, ok := m[id]
+	return idx, ok
+}
+
+// CommitRange is a tuple of commit indexes, one of which is known to be
+// committed, and the other one being the highest one that could potentially
+// be committed as more voters report back.
+type CommitRange struct {
+	Definitely uint64
+	Maybe      uint64
```

Comment:
In the commit index case we don't care because we always have all voters report in (and even if we didn't have that, we'd probably still not care). But in the voter case this matters because only when `Definitely == Maybe` is the vote over. Added a comment.

Reply:
See the above comment (which maybe wasn't visible to you when you wrote the review). I also added similar verbiage inline.

Reply:
If `Maybe` is just for votes, it feels like we're really stretching the commit index abstraction. We map yes/no votes to numbers so we treat votes as commit indexes, but then we have to add extra complexity to the commit index handling. Maybe it's better to just implement voting and index committing separately.

Reply:
> If Maybe is just for votes, it feels like we're really stretching the commit index abstraction. 

Are we stretching the abstraction? It just so happens that the one caller that wants a committed index always passes acks (even if they might be zero) for all members (and so `Definitely == Maybe` is guaranteed). The library is more "powerful" than it needs to be for the commit index case, but that's OK. Mapping the process of voting into a committed index computation is straightforward (at least as I perceive it).

Btw, to give a really clean example of where this matters, consider a config with two nodes in which one replica has voted "no" and the other one hasn't voted yet. The result is `VoteLost`, but nobody has reached quorum. What matters is that "yes" cannot reach quorum any more. So we need some code that tells you that.

Btw, sorry for going back and forth on this and not just "doing it". These are small things but I've found in the past that there's something to learn from them. In that vein, if you're still uncomfortable with this, please give some more detail on why.

Reply:
> It just so happens that the one caller that wants a committed index always passes acks (even if they might be zero) for all members (and so Definitely == Maybe is guaranteed)

This is where the semantics get muddled. You're treating the passed-in ack as a certainty and getting Definitely == Maybe, while I would say that the passed-in acks are lower bounds (with uncertainty in one direction), so Maybe is always maxint (I didn't realize that you'd get Maybe == Definitely until this comment). Indexes can always increase; there's never a case in which an entry becomes impossible to commit (except when we lose our leadership and have to step down to follower). 

Votes work differently: uncertainty is a third state and you can't transition from one certain state to another. I worry that even though there are some similarities between the append and vote cases, trying to handle them both in one interface invites confusion. I know there are tests and no one's going to change this code lightly, but the way it's organized it seems like it would be attractive to try and change the implementation in a way that only makes sense in one use case.

Reply:
The confusion around `CommittedIndex` is understandable. I don't share it but both you and Nathan have had it so it's definitely objectively confusing. Thanks for pushing me on this, I've updated the branch to add a simple implementation for voting and removed CommitRange and all of its vestiges. We lose a little bit of test coverage but it was abundant to begin with and still is, so I'm not worried about that.